File: bbsync/api/v1/activity/Activity.js
// The MIT License (MIT)

// Copyright (c) 2015 Elliott Richerson

// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:

// The above copyright notice and this permission notice shall be included in 
// all copies or substantial portions of the Software.

// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

module.exports = function Activity(REST, db, validate, errors) {

    var schema = [{
        attribute: "name",
        type: "String",
        required: true,
        auto: false,
        test: validate.activity_name()
    }, {
        attribute: "icon",
        type: "String",
        required: true,
        auto: false,
        test: validate.activity_icon()
    }, {
        attribute: "warn",
        type: "Double",
        required: true,
        auto: false,
        test: validate.activity_warn()
    }, {
        attribute: "critical",
        type: "Double",
        required: true,
        auto: false,
        test: validate.activity_critical()
    }, {
        attribute: "created_on",
        type: "Date",
        required: false,
        auto: true
    }, {
        attribute: "updated_on",
        type: "Date",
        required: false,
        auto: true
    }];

    var fxns = {
        post: db.activity_create.bind(db),
        getOne: db.activity_by_id.bind(db),
        getAll: db.activities_all.bind(db),
        put: db.activity_update.bind(db),
        del: db.activity_delete_by_id.bind(db)
    };
    var activityREST = new REST(fxns, schema, validate, errors, "Activity");

    var activity = {
        schema: schema,
        post: activityREST.post,
        get: activityREST.get,
        put: activityREST.put,
        del: activityREST.del
    }
    return activity;
};

File: bbsync/mobile/iOS/Baby Sync/Baby Sync/Activity.swift
//
//  Activity.swift
//  Baby Sync
//
//  Created by Elliott Richerson on 9/29/15.
//  Copyright © 2015 Elliott Richerson. All rights reserved.
//

import Foundation

struct Activity {
    var id: Int = 0
    var name: String = ""
    var icon: String = ""
    var warn: TimeInterval = TimeInterval()
    var critical: TimeInterval = TimeInterval()
    var createdOn: Date = Date()
    var updatedOn: Date = Date()
    
    init() {
        self.id = 0
        self.name = ""
        self.icon = ""
        self.warn = 0.0
        self.critical = 0.0
        self.createdOn = Date()
        self.updatedOn = Date()
    }
    
    init(activity: JSON) {
        self.id = activity["id"].intValue
        self.name = activity["name"].stringValue
        self.icon = activity["icon"].stringValue
        self.warn = activity["warn"].doubleValue
        self.critical = activity["critical"].doubleValue
        self.createdOn = ISO8601DateFormatter.sharedInstance.date(from: activity["created_on"].stringValue)!
        self.updatedOn = ISO8601DateFormatter.sharedInstance.date(from: activity["updated_on"].stringValue)!
    }
    
    init(activity: Activity) {
        self.id = activity.id
        self.name = activity.name
        self.icon = activity.icon
        self.warn = activity.warn
        self.critical = activity.critical
        self.createdOn = activity.createdOn
        self.updatedOn = activity.updatedOn
    }
}
